src/charm.py

```diff
@@ -57,6 +59,12 @@ def __init__(self, *args):
         self.password_action_events = PasswordActionEvents(self)
         self.tls_events = TLSEvents(self)
         self.provider_events = ProviderEvents(self)
+        self.upgrade_events = ZKUpgradeEvents(
+            self,
+            dependency_model=ZooKeeperDependencyModel(
+                **DEPENDENCIES  # pyright: ignore[reportGeneralTypeIssues]
+            ),
+        )
```

Comment:
During pod failover/rescheduling, we also get `upgrade-charm`.
During upgrades on K8s, the FIRST event we get is `upgrade-charm`.

We want the pod to failover normally, so we handle `charm._on_zookeeper_pebble_ready`.
However, we want to block this during an 'actual' upgrade upgrade. As such, we check for the existence of an `upgrade_stack` (which will have been set if we ran `pre-upgrade-check`). If so, the charm skips `charm._on_zookeeper_pebble_ready`, but DOES run `charm.upgrade_events._on_zookeeper_pebble_ready_upgrade` instead. 

We can observe to the same event multiple times, we skip the normal one and run the upgrade one in that case.

Reply:
ah I see. I find it a bit confusing that on failover/rescheduling we run this event, but all right, that's the framework. Now I understand why we do this! 

I'm just wondering whether it is a good pattern to observe an event multiple times? I'm wondering whether we could centralize this logic in ONE upgrade method. Just saying this, because I would believe it is easier to maintain a code if we know that an event is processes in one place only, rather than having the possibility of the code related to one event to be scattered around

Reply:
So I agree on your last point. I think it's a bit funky and (apparently) confusing.

The lib should be looked at for that, but definitely with more attention :)